### PR TITLE
[algo] feat: add CPPO (position-weighted cumulative-prefix-divergence token mask)

### PR DESCRIPTION
**Pull request by chongqichuizi875**
### What does this PR do?

Adds the official implementation of **CPPO** (Cumulative Prefix-divergence Policy
Optimization), from the paper *"Beyond Uniform Token-Level Trust Region in LLM
Reinforcement Learning"* (arXiv:2606.10968).

A uniform per-token trust region is a poor fit for autoregressive generation: an early-token
deviation cascades through the entire suffix, and a pointwise threshold ignores how far the
prefix has already drifted from the rollout policy. CPPO keeps the token-level
ratio-advantage surrogate and the same per-token divergence `D_t = |π(y_t|s_t) − μ(y_t|s_t)|`,
but reallocates the divergence budget along the response via two coupled mechanisms:

1. a **position-weighted** token-level threshold (a decreasing weight `w_t ∈ [w_min, 1]`), so
   early tokens are constrained more tightly than late ones, and
2. a **cumulative prefix-average budget** `δ_b`, so the threshold tightens as off-policy
   prefix drift accumulates.

Only the masking decision changes; the surrogate objective is unchanged (no new loss terms).
This PR ships the self-contained **Binary-TV** variant of the divergence.

- Paper: https://arxiv.org/pdf/2606.10968
- Project page: https://hunyuan-cppo.github.io

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here:
  https://github.com/volcengine/verl/pulls?q=is%3Apr+cppo
- [x] Format the PR title as `[{modules}] {type}: {description}` — `[algo] feat: add CPPO ...`.

### Test

CPPO is an algorithm/loss change that CI cannot validate, so it was validated by experiment.

Experiments on **Qwen3-30B-A3B-Base** (DAPO-Math train, AIME 24/25/26 eval, 16k rollout),
following the paper's setting (δ=0.20, w_min=0.8, δ_b=0.02, train_batch=256, rollout.n=16):

- **Avg@16** — CPPO reaches **54.79%** on AIME 24/25/26.
image
- **Reward (train)** — reward rises smoothly without late-stage collapse.
image
- **Training stability** — the training–inference probability mismatch stays controlled
  throughout training (no divergence blow-up).
image

The mask math was additionally cross-checked offline against a scalar reference
(vectorized cumsum vs. a per-token loop) over five edge-case scenarios.

### API and Usage Example

CPPO adds two config fields on top of the shared token-level threshold scale
(`actor.clip_ratio`); no API breakage. Selected via `actor.policy_loss.loss_mode=cppo`.

```bash
# Qwen3-30B-A3B-Base, Binary-TV
bash examples/cppo_trainer/run_qwen3_30b_a3b_megatron.sh
```

```yaml
actor_rollout_ref.actor.policy_loss.loss_mode: cppo
actor_rollout_ref.actor.clip_ratio: 0.20                 # token-level divergence threshold δ
actor_rollout_ref.actor.policy_loss.cppo_w_min: 0.8      # position-weight floor, w_t ∈ [w_min, 1]
actor_rollout_ref.actor.policy_loss.cppo_delta_b: 0.02   # prefix-average budget floor δ_b
```

CPPO's divergence is measured against the **rollout** policy μ, so `old_log_probs` must be
the rollout log-probs. The example runs the synchronous trainer
(`verl.trainer.main_ppo_sync`) with `algorithm.rollout_correction.bypass_mode=True`, which
sets `old_log_probs = rollout_log_probs` and leaves `loss_mode` untouched (requires
`pip install TransferQueue`).

### Design & Code Changes

- `verl/trainer/ppo/core_algos.py` — new `@register_policy_loss("cppo")`
  `compute_policy_loss_cppo`: Binary-TV per-token divergence, position-weighted threshold,
  and a cumulative prefix budget with a per-sequence dynamic
  `δ_b = clamp(P90(D_t), δ_b, 5·δ_b)`. The mask is computed under `no_grad` (it is a
  trust-region gate); the returned metrics match the existing divergence-loss metrics
  (`pg_clipfrac` / `ppo_kl` / `pg_clipfrac_lower`).
- `verl/workers/config/actor.py` — two new `PolicyLossConfig` fields (`cppo_w_min=0.8`,
  `cppo_delta_b=0.02`); δ reuses the existing `clip_ratio` field.
- `verl/trainer/config/actor/actor.yaml` + the four `_generated_*.yaml` — config field
  declarations plus the `autogen-trainer-cfg` regeneration.
- `examples/cppo_trainer/` — `run_qwen3_30b_a3b_megatron.sh` + `README.md`.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [x] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: CPPO is an algorithm/loss change; correctness is validated by the experiments above (CI has no GPU to run an RL training step), mirroring how other policy-loss variants are merged.
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ).
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by chongqichuizi875** [comment]
Thanks @gemini-code-assist! Addressed in the latest push:
  - ✅  Replaced the NaN `torch.where` with `torch.nan_to_num(q_seq, nan=delta_b)`.
  - ✅  Replaced `torch.minimum(torch.full_like(...), ...)` with `torch.clamp(..., max=delta)`.
  - Kept `assert config is not None` / `assert not isinstance(config, AlgoConfig)`: these are the
    established convention across all policy-loss functions in this file (15 and 7 occurrences),
    so removing them only here would be inconsistent.

File: verl/trainer/config/actor/actor.yaml [hunk 1]
```diff
@@ -75,6 +75,18 @@ policy_loss:
   # KL divergence penalty coefficient
   ppo_kl_coef: 0.1
 
+  # CPPO (cppo.pdf). Only consumed when loss_mode == "cppo". The token-level threshold
```
Comment (wuxibin89) [inline verl/trainer/config/actor/actor.yaml]: Could you group cppo specific parameter as below?
```yaml
cppo:
  cppo_w_min: 0.8
  cppo_delta_b: 0.02
  cppo_delta_b_q: 0.9
  cppo_delta_b_k: 1.0
```

Comment (chongqichuizi875) [reply verl/trainer/config/actor/actor.yaml]: Thanks @wuxibin89! Done in the latest push (aaabbbb):

  - ✅  Grouped all CPPO-specific parameters under a nested `cppo:` section, as suggested.
    They now live in a dedicated `CPPOConfig` on `policy_loss.cppo`, e.g.
    `actor.policy_loss.cppo.cppo_w_min=0.8`. Updated the dataclass, `actor.yaml`, the
    example script/README, and regenerated the `_generated_*.yaml` configs accordingly.

  I kept the `cppo_` prefix on the field names (so it reads `cppo.cppo_w_min` etc.),
  matching how you wrote it in the comment. Happy to drop the prefix to plain
  `cppo.w_min` if you'd prefer the shorter form — just let me know.

**Comment by chongqichuizi875** [comment]
We have tested our demo script. Here is the [wandb link ](https://wandb.ai/sumailmao-tencent/demo-for-verl-pr?nw=nwuserchongqichuizi)
@wuxibin89 
image
